🎉 Narrative charts

sophiamersmann opened the pull request:
resolves #2215 

### Summary

- allow gdoc authors to overwrite the title and/or subtitle of a chart
- if the title or subtitle is overwritten, Grapher renders a minimal view hiding all controls and all tabs (expect for the sources tab)
- authors can add back controls and tabs one by one (or, if needed, add back all controls and all tabs)

<details>
<summary><b>Example with screenshots</b></summary>

Archie block and chart:

```
{.chart}
url: https://ourworldindata.org/grapher/countries-democracies-autocracies-row?stackMode=absolute
title: The number of democracies has declined in recent years
{}
```

<img width="853" alt="Screenshot 2023-06-01 at 10 55 14" src="https://github.com/owid/owid-grapher/assets/12461810/4f3215aa-cd85-41d8-9bfa-1c88695d1e4d">

The original chart looks like:

<img width="748" alt="Screenshot 2023-05-30 at 11 34 37" src="https://github.com/owid/owid-grapher/assets/12461810/588afa61-e9b3-4897-852f-0dafe132a5a1">

</details>

<details>
<summary><b>Example with screenshots</b></summary>

Archie block and chart:

```
{.chart}

url: https://ourworldindata.org/grapher/dietary-choices-of-british-adults

-- overwrite title and subtitle
title: Under-50s are more likely to report being vegetarian, vegan, or pescetarian
subtitle: Self-reported dietary choices of Brits (YouGov, Jan 2023)

-- show individual tabs
[.tabs]
- table
- download
[]

{}
```

<img width="853" alt="Screenshot 2023-06-01 at 10 55 40" src="https://github.com/owid/owid-grapher/assets/12461810/42ad267c-6c51-4587-8f2e-b06172bc9774">

The original chart looks like:

<img width="746" alt="Screenshot 2023-05-30 at 12 04 03" src="https://github.com/owid/owid-grapher/assets/12461810/896f4599-4072-4b61-996d-17cefd825ec2">

</details>

<details>
<summary><b>Example with screenshots</b></summary>

Archie block and chart:

```
{.chart}
url: https://ourworldindata.org/grapher/national-poverty-line-vs-gdp-per-capita?zoomToSelection=true&country=DEU~ARG~AUS~AUT~BEL~CAN~CHL~HRV~CYP~CZE~DNK~EST~FIN~FRA~GRC~HUN~ISL~IRL~ISR~ITA~LVA~LTU~LUX~MLT~NLD~NOR~POL~PRT~SYC~SVK~SVN~KOR~ESP~SWE~CHE~GBR~USA~URY
title: Some interesting fact about high-income countries

[.controls]
- zoomToggle
[]
{}
```

<img width="852" alt="Screenshot 2023-06-02 at 10 47 22" src="https://github.com/owid/owid-grapher/assets/12461810/50972193-767e-42bc-bf14-b9beb6f3050a">

The original chart looks like:

<img width="747" alt="Screenshot 2023-06-02 at 12 00 18" src="https://github.com/owid/owid-grapher/assets/12461810/0b6e0b70-2b78-458f-86b8-960c03052b20">

</details>

### Technical details

- The Grapher's `<figure />` element gets an additional attribute `data-grapher-config` that is either unset or a valid Grapher config
- When the Grapher chart is rendered, the config loaded from the grapher page is merged with the additional "local" config that comes from the `<figure />` element

**Properties added to the Chart block:**
- `title`: overwrite title (triggers minimal view)
- `subtitle`: overwrite subtitle (triggers minimal view)
- `controls`: list of controls to show (minimal view renders none by default); valid values include:
    - `all`: shortcut to render all controls of the original chart
    - `relativeToggle`
    - `timeline`
    - `facetControl`
    - `entitySelector`
    - `zoomToggle`
    - `noDataAreaToggle`
    - `alignAxisScalesToggle`
    - `xLogLinearSelector`
    - `yLogLinearSelector`
- `tabs`: list of tabs to show (minimal view renders only the sources tab and the default chart/map/table tab); valid values include:
    - `all`: shortcut to render all tabs of the original chart
    - `chart`
    - `map`
    - `table`
    - `download`

The chart/map/table tabs are configured automatically which means that setting the chart/map/table tab manually in a gdocs block is only necessary when the author wishes to show that tab in addition to the automatically configured/default one (note that having the table tab as initial view is rare but possible)

### To do

- [x] Update Gdocs documentation

From discussions:

- [x] Keep sources tab
- [x] Remove related question

sophiamersmann commented:
cc @ikesau no review required (unless you want to) but I wanted to ping you in case you wanted to have a look at the gdocs code

sophiamersmann commented on packages/@ourworldindata/utils/src/owidTypes.ts:
```diff
@@ -510,6 +510,32 @@ export type EnrichedBlockAside = {
     caption: Span[]
 } & EnrichedBlockWithParseErrors
 
+export const validChartShowKeywords = [
+    "all tabs",
+    "all controls",
+    "relative toggle",
+    "timeline",
+    "facet control",
+    "entity selector",
+    "country selector",
+    "zoom toggle",
+    "no data area toggle",
+    "data area toggle",
+    "area toggle",
+    "align axis scales toggle",
+    "align axis toggle",
+    "x log/linear selector",
+    "x log selector",
+    "y log/linear selector",
+    "y log selector",
+    "chart tab",
+    "map tab",
+    "table tab",
+    "download tab",
+] as const
```
yeah, probably good to split them up!

danyx23 replied:
I'm a bit torn on this - I find the separation clearer if tabs are separate but it might be that very soon we'll change the visual UI elements of grapher so that the things that are currently all tabs are no longer one such clearly delinieated group. It probably doesn't matter too much though so I'm fine with both

sophiamersmann replied:
Regarding the format: In the spirit of GDocs and Archie, I wanted to give it a natural language feel. I agree though that, as a developer, kebap-casing would also make me feel more comfortable. I guess what matters in the end is a certain consistency across gdoc blocks, so maybe @ikesau or @danyx23 have thoughts?

sophiamersmann replied:
Oh, I didn't see your comment, Daniel, when I posted mine!

sophiamersmann replied:
I just changed `show` to `tabs` and `controls` – if you're fine with both, I'll keep them separated

sophiamersmann commented on site/gdocs/Chart.tsx:
```diff
@@ -42,3 +89,57 @@ export default function Chart({
     )
 }
+
+const mapKeywordToGrapherConfig = (
```
yeah I wanted it to be forgiving but I think you're right that this will probably lead to more confusion / headache in the future... I'll remove them!

danyx23 replied:
I agree with Christian here - I think the ambiguity here is not particularly helpful. In the end, this is a controlled vocabulary that matches onto something pretty technical. At the same time, the fields that we want to influence have a different origin story and so some are called hide* and some are called has* etc.

We use this inside a key called `show` - as such I'd be in favour of using a camelCase fragment that matches the config field this relates to. This should make it straightforward to figure out the field this actually influences. Authors are exposed to these names already through tools like the bulk chart editor.

So I'd suggest names like
```
relativeToggle
timeline
facetControl
chartTab
```

danyx23 replied:
Ah you posted your comment just as I had finished writing it as part of my review :)

sophiamersmann replied:
haha this is all a bit confusing

Good point that authors are exposed to these names anyway. I'll camelCase them!

danyx23 commented on site/multiembedder/MultiEmbedder.tsx:
```diff
@@ -175,16 +178,38 @@ class MultiEmbedder {
+                ? JSON.parse(figureConfigAttr)
+                : {}
+
+            // make sure the tab of the active pane is visible
+            if (figureConfigAttr) {
+                const activeTab =
+                    queryParams.tab ||
+                    grapherPageConfig.tab ||
+                    GrapherTabOption.chart
+                if (activeTab === GrapherTabOption.chart)
+                    localConfig.hasChartTab = true
+                if (activeTab === GrapherTabOption.map)
+                    localConfig.hasMapTab = true
+                if (activeTab === GrapherTabOption.table)
+                    localConfig.hasTableTab = true
+            }
```
I kinda like that it is done here explicitly

danyx23 reviewed (commented):
Great work! I'll still play with this a little bit. I'm in favour of renaming the show field values but otherwise this looks good. Ping me once more when the field values are changed, then I should be done with my live testing as well.

sophiamersmann commented:
Thank you for your reviews, @samizdatco and @danyx23! I think the API is much nicer now :)

The PR's history is a bit confusing because I pushed changes and commented at the exact time Daniel reviewed. This is what I changed:
- split `show` into `controls` and `tabs`
- removed duplicate keywords for an option
- camelCased options

PS: I updated the description accordingly.

danyx23 reviewed (approved):
Looks good!
